audit/client.go

```diff
@@ -0,0 +1,231 @@
+// Package pki provides support for HSDP CDR service
+//
+// We only intent to support the CDR FHIR STU3 and newer with this library.
+package audit
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"strings"
+
+	"go.elastic.co/apm/module/apmhttp"
+
+	signer "github.com/philips-software/go-hsdp-signer"
+
+	"github.com/google/fhir/go/jsonformat"
+)
+
+const (
+	libraryVersion = "0.29.0"
+	userAgent      = "go-hsdp-api/audit/" + libraryVersion
+	APIVersion     = "2"
+)
+
+// OptionFunc is the function signature function for options
+type OptionFunc func(*http.Request) error
+
+// Config contains the configuration of a client
+type Config struct {
+	Region       string
+	Environment  string
+	RootOrgID    string
+	AuditBaseURL string
+	SharedKey    string
+	SharedSecret string
+	TimeZone     string
+	DebugLog     string
+}
+
+// A Client manages communication with HSDP CDR API
+type Client struct {
+	config        *Config
+	httpClient    *http.Client
+	auditStoreURL *url.URL
+
+	// User agent used when communicating with the HSDP IAM API.
+	UserAgent string
+
+	ma         *jsonformat.Marshaller
+	um         *jsonformat.Unmarshaller
+	httpSigner *signer.Signer
+
+	debugFile *os.File
+}
+
+// NewClient returns a new HSDP Audit API client. Configured console and IAM clients
+// must be provided as the underlying API requires tokens from respective services
+func NewClient(httpClient *http.Client, config *Config) (*Client, error) {
+	return newClient(httpClient, config)
+}
+
+func newClient(httpClient *http.Client, config *Config) (*Client, error) {
+	var err error
+
+	if httpClient == nil {
+		httpClient = apmhttp.WrapClient(http.DefaultClient)
+	}
+
+	c := &Client{httpClient: httpClient, config: config, UserAgent: userAgent}
+	if config.DebugLog != "" {
+		var err error
+		c.debugFile, err = os.OpenFile(config.DebugLog, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
+		if err != nil {
+			c.debugFile = nil
+		}
+	}
+	c.httpSigner, err = signer.New(c.config.SharedKey, c.config.SharedSecret)
+	if err != nil {
+		return nil, fmt.Errorf("signer.New: %w", err)
+	}
+	ma, err := jsonformat.NewMarshaller(false, "", "", jsonformat.STU3)
```

Comment:
HSDP Audit is DSTU2... is it ok to use STU3 to unmarshal?

Reply:
Yes, apparently. I created a [PR](https://github.com/google/fhir/pull/31) for the Google FHIR repo but that might take weeks / months / if at all to get merged. Based on the changes I made in the PR I suspected that at least for Marshalling to JSON the `STU3` marshaller should work just as well on `DSTU2` proto messages. So far this looks to be the case